File: src/ell_object.rs
//! Elliptic Geometry
//!
//! This module provides support for elliptic geometry using the Cayley-Klein
//! model. In elliptic geometry, the perpendicular relationship is defined by the
//! polarity induced by the elliptic quadric, where each point is self-dual.

use crate::pg_object::{EllipticLine, EllipticPoint};
use crate::{CayleyKleinPlane, CayleyKleinPlanePrimitive};

impl_cayley_klein_plane!(
    EllipticPoint,
    EllipticLine,
    |p: &EllipticPoint| EllipticLine::new(p.coord),
    |l: &EllipticLine| EllipticPoint::new(l.coord)
);

File: python_ai/src/projgeom/ell_object.py
"""Elliptic Geometry Objects module.

This module defines elliptic geometry points and lines with their
perpendicularity operations.
"""

from .ck_plane import CayleyKleinPlane
from .pg_object import EllipticLine, EllipticPoint


def _perp_point(point: EllipticPoint) -> EllipticLine:
    """Return the polar line of an elliptic point."""
    return EllipticLine(point.coord.copy())


def _perp_line(line: EllipticLine) -> EllipticPoint:
    """Return the pole of an elliptic line."""
    return EllipticPoint(line.coord.copy())


# Implement CayleyKleinPlanePrimitive for EllipticPoint
EllipticPoint.perp = _perp_point  # type: ignore[attr-defined]

# Implement CayleyKleinPlanePrimitive for EllipticLine
EllipticLine.perp = _perp_line  # type: ignore[attr-defined]

# Implement CayleyKleinPlane for EllipticPoint and EllipticLine
CayleyKleinPlane.register(EllipticPoint)
CayleyKleinPlane.register(EllipticLine)
